src/persist/src/nemesis/generator.rs

```diff
@@ -7,89 +7,689 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use rand::prelude::SmallRng;
+use std::collections::HashMap;
+
+use rand::prelude::{IteratorRandom, SliceRandom, SmallRng};
 use rand::{Rng, SeedableRng};
+use serde::Serialize;
 
 use crate::nemesis::{
     Input, ReadSnapshotReq, Req, ReqId, SealReq, SnapshotId, TakeSnapshotReq, WriteReq,
 };
 
+/// Configuration of the relative probabilities of producing various
+// operations.
+#[derive(Clone, Debug, Serialize)]
+pub struct GeneratorConfig {
+    pub write_unsealed_weight: u32,
+    pub write_sealed_weight: u32,
+    pub seal_weight: u32,
+    pub take_snapshot_weight: u32,
+    pub read_snapshot_weight: u32,
+}
+
+impl GeneratorConfig {
+    fn all_operations() -> GeneratorConfig {
+        GeneratorConfig {
+            write_unsealed_weight: 1,
+            write_sealed_weight: 1,
+            seal_weight: 1,
+            take_snapshot_weight: 1,
+            read_snapshot_weight: 1,
+        }
+    }
+}
+
+impl Default for GeneratorConfig {
+    fn default() -> Self {
+        Self::all_operations()
+    }
+}
+
+struct GeneratorState {
+    sealed_by_stream: HashMap<String, u64>,
+    snap_id: SnapshotId,
+    outstanding_snaps: HashMap<SnapshotId, String>,
+}
+
+impl Default for GeneratorState {
+    fn default() -> Self {
+        GeneratorState {
+            sealed_by_stream: HashMap::new(),
+            snap_id: SnapshotId(0),
+            outstanding_snaps: HashMap::new(),
+        }
+    }
+}
+
+enum ReqGenerator {
+    WriteUnsealed,
+    WriteSealed,
+    Seal,
+    TakeSnapshot,
+    ReadSnapshot,
+}
+
+fn rng_stream(rng: &mut SmallRng) -> String {
+    // TODO: This is meant to be Zipf, use it once rust-random/rand/1136 lands.
+    rng.gen_range('a'..'e').to_string()
```

Comment:
Ha! You have no idea how much time I spent last Friday trying to figure out how to do this without WeightedIndex. From a practical perspective, you're probably right, but I'm bothered that I can't just do this in terms of some generic impl of Distribution that defaults to Zipf.

kvnemesis breaks this up into two input types: one for existing keys and one for new keys. I think something similar would be useful for us b/c compactions. (I wouldn't split up WriteSealed since that just always errors, so we'd end up with something like WriteNewKey, WriteExistingKey, and WriteSealed.) Thinking out loud, your approach would require that we recompute the WeightedIndex each time we mint a new key, which seems totally fine. Yup, you're totally right.

Reply:
I have a very slight preference to avoid fancy distributions because they make it harder for me to prove simple properties. Like if I wanted to validate "I'm going to write to 7 keys and 1 key will be responsible for 99% of writes" then I have to look at the zipfian distribution wikipedia page to verify that I set the s parameter correctly to satisfy the desired property, or alternatively, can verify at a glance with a set of weights.

Reply:
Yeah that's fair! If we do add a variable (and unbounded) number of keys/streams, we'll need to be able to generate the weights programmatically though. Thoughts?

Reply:
I think a structure like "this subset H of items is hot and accounts for X% of <operations> sampled uniformly from H, and then the remain (100 - X)% of items are sampled uniformly from all items"

is a relatively simple structure that works for most workloads (and the empty version of this is nicely uniform), and can accommodate changing sets of keys pretty easily without having to recompute a large set of weights. I also think its legit once we have a huge set of keys / streams to fall back to a zipfian distribution